core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleDriver.java

```diff
@@ -937,7 +937,7 @@ private void derive() {
                 subset.disableEnforcing();
               }
               RelSubset relSubset = planner.register(newRel, rel);
-              assert relSubset.set == planner.getSubsetNonNull(rel).set;
+              assert relSubset.set == VolcanoPlanner.canonize(planner.getSubsetNonNull(rel)).set;
```

Comment:
This looks really suspicious.
Should subset merge cleanup the references? Should `getSubsetNonNull` return the canonical subset?

Reply:
Agree with @vlsi.